Remove unsupported codecs from hwdec-codecs

low-batt opened the pull request:
This commit will:
- Add a new class `HardwareDecodeCapabilities`
- Add a call to `HardwareDecodeCapabilities.checkCapabilities` in `AppDelegate.applicationWillFinishLaunching`
- Add a new method `userOptionsContains` to `MPVController` that checks if a user has configured a `mpv` option in IINA's advanced settings
- Add a new property `mpvCodecToCodecTypes` to  `MPVController` that maps a `mpv` codec name to core media video codec types
- Add a new `adjustCodecWhiteList` method to `MPVController` that removes codecs from the `mpv` `hwdec-codecs` option if the Mac does not support hardware decoding for that codec
- Change the `applyHardwareAccelerationWorkaround` method to use the value of the `hwdec-codecs` option instead of a hardcoded value

These changes cause mpv to not attempt to use hardware decoding for a codec that is known to not be supported. This eliminates the overhead of setting up for hardware decoding only to have it fail. This is not comprehensive. This only covers the recent codecs whose support for hardware decoding varies among Macs. This merely reduces the dependence upon the FFmpeg fallback to software decoding feature in some cases.

- [x] I have read [CONTRIBUTING.md](https://github.com/iina/iina/blob/develop/CONTRIBUTING.md)
- [ ] This implements/fixes issue #.

---

**Description:**

low-batt committed:
Remove unsupported codecs from hwdec-codecs

This commit will:
- Add a new class HardwareDecodeCapabilities
- Add a call to HardwareDecodeCapabilities.checkCapabilities in
  AppDelegate.applicationWillFinishLaunching
- Add a new method userOptionsContains to MPVController that checks if a
  user has configured a mpv option in IINA's advanced settings
- Add a new property mpvCodecToCodecTypes to  MPVController that maps a
  mpv codec name to core media video codec types
- Add a new adjustCodecWhiteList method to MPVController that removes
  codecs from the mpv hwdec-codecs option if the Mac does not support
  hardware decoding for that codec
- Change the applyHardwareAccelerationWorkaround method to use the value
  of the hwdec-codecs option instead of a hardcoded value

These changes cause mpv to not attempt to use hardware decoding for a
codec that is known to not be supported. This eliminates the overhead of
setting up for hardware decoding only to have it fail. This is not
comprehensive. This only covers the recent codecs whose support for
hardware decoding varies among Macs. This merely reduces the dependence
upon the FFmpeg fallback to software decoding feature in some cases.

low-batt commented:
Rebased with `develop` and fixed merge conflicts.

uiryuu commented on iina/HardwareDecodeCapabilities.swift:
```diff
@@ -0,0 +1,103 @@
+    if #available(macOS 11.0, *) {
+      VTRegisterSupplementalVideoDecoderIfAvailable(codecType)
+    }
+    if #available(macOS 10.13, *) {
```
Given that IINA now will only run on macOS 10.15+, we can safely remove this version check and the code after this, since they will never be executed.

low-batt replied:
Fixed

low-batt commented:
@lhc70000 Another PR that is affected by the commit that was amended after it was merged into `develop`. Please fix.

low-batt committed:
Remove unsupported codecs from hwdec-codecs

This commit will:
- Add a new class HardwareDecodeCapabilities
- Add a call to HardwareDecodeCapabilities.checkCapabilities in
  AppDelegate.applicationWillFinishLaunching
- Add a new method userOptionsContains to MPVController that checks if a
  user has configured a mpv option in IINA's advanced settings
- Add a new property mpvCodecToCodecTypes to  MPVController that maps a
  mpv codec name to core media video codec types
- Add a new adjustCodecWhiteList method to MPVController that removes
  codecs from the mpv hwdec-codecs option if the Mac does not support
  hardware decoding for that codec
- Change the applyHardwareAccelerationWorkaround method to use the value
  of the hwdec-codecs option instead of a hardcoded value

These changes cause mpv to not attempt to use hardware decoding for a
codec that is known to not be supported. This eliminates the overhead of
setting up for hardware decoding only to have it fail. This is not
comprehensive. This only covers the recent codecs whose support for
hardware decoding varies among Macs. This merely reduces the dependence
upon the FFmpeg fallback to software decoding feature in some cases.
